core/src/server/subscription.rs

```diff
@@ -0,0 +1,486 @@
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Subscription related types and traits for server implementations.
+
+use super::helpers::{MethodResponse, MethodSink};
+use crate::server::error::{DisconnectError, PendingSubscriptionAcceptError, SendTimeoutError, TrySendError};
+use crate::server::rpc_module::ConnectionId;
+use crate::{traits::IdProvider, Error};
+use jsonrpsee_types::{response::SubscriptionError, ErrorObjectOwned, Id, SubscriptionId, SubscriptionResponse};
+use parking_lot::Mutex;
+use rustc_hash::FxHashMap;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{sync::Arc, time::Duration};
+use tokio::sync::{mpsc, OwnedSemaphorePermit, Semaphore};
+
+/// Type-alias for subscribers.
+pub type Subscribers = Arc<Mutex<FxHashMap<SubscriptionKey, (MethodSink, mpsc::Receiver<()>)>>>;
+/// Subscription permit
+pub type SubscriptionPermit = OwnedSemaphorePermit;
+
+/// Convert something into a subscription close notification
+/// before a subscription is terminated.
+pub trait IntoSubscriptionCloseResponse {
+	/// Convert something into a subscription response
+	fn into_response(self) -> SubscriptionCloseResponse;
+}
+
+/// Represents what action that will sent when a subscription callback returns.
+#[derive(Debug)]
+pub enum SubscriptionCloseResponse {
+	/// No further message will be sent.
+	None,
+	/// Send a subscription notification.
+	///
+	/// The subscription notification has the following format:
+	///
+	/// ```json
+	/// {
+	///  "jsonrpc": "2.0",
+	///  "method": "<method>",
+	///  "params": {
+	///    "subscription": "<subscriptionID>",
+	///    "result": <your msg>
+	///    }
+	///  }
+	/// }
+	/// ```
+	Notif(SubscriptionMessage),
+	/// Send a subscription error notification
+	///
+	/// The error notification has the following format:
+	///
+	/// ```json
+	/// {
+	///  "jsonrpc": "2.0",
+	///  "method": "<method>",
+	///  "params": {
+	///    "subscription": "<subscriptionID>",
+	///    "error": <your msg>
+	///    }
+	///  }
+	/// }
+	/// ```
+	NotifErr(SubscriptionMessage),
+}
+
+impl<T, E> IntoSubscriptionCloseResponse for Result<T, E>
```

Comment:
I think we `()` will be serialized as `null` which is annoying

Reply:
we could also implement that for `Result<(), Error>` instead

Reply:
As we said in the call, that sounds like a good idea to me!

I even wonder whether it makes sense for `Error` to be a concrete type (`?` will ".into()" other error types into it if possible anyway). This would remove the need to specify any generic types anywhere :)

Reply:
What I had in mind with this is that you can also inject your own JSON here such as

```
result: { "my": 120, "hehe": { "1":1} }
```

if we put `Error` there that is not possible it would just be:

```
result: "<err.to_string()>"
```

however, `ToString` here doesn't work super well either I think we would add extra string quotes ("")

Reply:
sorry, I didn't realize that I wrote `Error` above I meant `E: ToString` :)

Reply:
> What I had in mind with this is that you can also inject your own JSON here such as
> 
> ```
> result: { "my": 120, "hehe": { "1":1} }
> ```
> 
> if we put `Error` there that is not possible it would just be:
> 
> ```
> result: "<err.to_string()>"
> ```
> 
> however, `ToString` here doesn't work super well either I think we would add extra string quotes ("") such if one does:
> 
> ```
> sink.send().await.map_err(|_| serde_json::to_string(&MyError).unwrap())?;
> ```
> 
> lemme create a test for that

`E` impl's `ToString` at the moment, but for instance if we had a concrete `StringError` type for instance that serialized to a string correctly, and add `impl <T: ToString> From<T> for StringError` then all `?`'s would continue to work (as long as the errors impl Display which they generally always do) and no generics would be needed.

Dunno; maybe that doesn't make sense?

Reply:
ok, it won't work with the proc macro code basically it's not possible to have the assumption that `?` will work it could be `()` or some other type.

Thus, I could just think of something like:

```rust
  if ret_ty == "Result" {
                #into_sub_response::into_response(context.as_ref().#rust_method_name(subscription_sink, #params_seq).await.map_err(Into::into))
  } else {
    #into_sub_response::into_response(context.as_ref().#rust_method_name(subscription_sink, #params_seq).await
  }
```

Reply:
Implemented something similar as we have for the client to check the return type ident but I'm not super happy with it.

It works but a little hacky but probably fine here.

Reply:
> ok, it won't work with the proc macro code basically it's not possible to have the assumption that ? will work it could be () or some other type.

Ah I didn't think it would impact the proc macro; I'd assuemd that any IntoSubscriptionCloseResponse thing would "just work" as a return type so it didn't matter. Just happens that for user facing code if they want to return a Result type then they can use "?"'s and it'll work out?

Reply:
fixed now